concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1154 @@
+- At the top of the pyramid, there are usually `fewer` tests because these tests take `longer`. However, these tests are a more realistic scenario of how the end user will use our product. E2E tests are an example of this.
+
+### Controller Integration Tests (kubebuilder)
+
+**Test Setup**
+Controllers using the kubebuilder framework bootstrap integration tests using `Ginkgo` (for testing) and `Gomega` (for assertions).
+Using the `controller-runtime/pkg/envtest` package, a kubernetes `control-plane` (`API server` and `etcd`) is started locally once before the first test runs (`BeforeSuite`).
+You can also use a custom Kubernetes cluster (see `USE_EXISTING_CLUSTER` environment variable).
+
+**Goal**
+The goal of a controller integration test is to test the Kubernetes controller in a `limited` Kubernetes `environment`. That means that dependencies of the controller - such as other controllers (APIRule controller) or external eventing system - are not present. These systems are only present in higher hierarchies of the testing pyramid.
+
+Controller integration tests should be reserved for cases when we care about the eventual state of the object or need a Kubernetes cluster.
+
+**See Also**
+- [kubebuilder documentation](https://book.kubebuilder.io/reference/envtest.html)
+- [envtest godoc](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest)
+
+### Unit Tests
+
+[Unit tests](https://en.wikipedia.org/wiki/Unit_testing) are used to test individual `units` of the code. To test the unit in an `isolated` fashion, external `dependencies` are typically `mocked`.
+Unit tests provide a very fast `feedback cycle` and help discovering bugs in a very `early stage`.
+The advantage of unit testing is that "[Code can be impossible or difficult to unit test if poorly written, thus unit testing can force developers to structure functions and objects in better ways.](https://en.wikipedia.org/wiki/Unit_testing)"
+The big disadvantage of unit tests is that they are tightly coupled to the code, so rewriting the code may require new or modified unit tests.
+
+## Testing Guidelines 
+
+The following section describes ugly tests or problems that we discovered in our code base. The testing guidelines provide suggestions how the tests can be written instead.
+The guides should be used whenever possible but are not set in stone. There might be cases where the proposed solution does not fit and that is ok.
+
+**Best Practices**:
+- Find a `balance` between [`KISS`](https://en.wikipedia.org/wiki/KISS_principle) (keep it simple stupid) and [`DRY`](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) (don't repeat yourself). Tests should be easily `understandable`. At the same time, you should `avoid` code `duplication` because it gets un`maintaineable` very fast.
```

Comment:
```suggestion
- Find a `balance` between [`KISS`](https://en.wikipedia.org/wiki/KISS_principle) (keep it simple stupid) and [`DRY`](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) (don't repeat yourself). Tests should be easily `understandable`. At the same time, you should `avoid` code `duplication` because it gets `unmaintainable` very fast.
```